RFD.md

```diff
@@ -0,0 +1,275 @@
+
+* All processes started by the server share the same process group, so that we could kill them all as a group when killing the command.
+
+#### Design details
+
+##### Log streaming
+
+When executing a new command, we will redirect its stdout/stderr to a command-specific log file. This makes it possible for us to stream log content to multiple clients at the same time using different stream positions and always have access to the whole command output no matter how long it is (limited by available disk space). The streaming of content from the end of the file may be implemented by using a library like [hpcloud/tail](https://github.com/hpcloud/tail) or built into the solution by either polling the `stat()` syscall results (and looking at the size of the file there) or by using the `inotify()` API for detecting file changes.
+
+There will not be any log rotation, but in a production scenario we would probably add a limit on the size of the log and start rotating it eventually.
+
+##### Process management
+
+We're going to use the `exec.Command` Go API to start each command. Linux kernel namespaces will be enabled for each command via the `SysProcAttr` attribute.  For filesystem isolation (`/proc remounting, etc) and for applying Cgroups to the command process, we're going to rely on the [reexec](https://pkg.go.dev/github.com/docker/docker/pkg/reexec) package from Docker to allow us to apply the changes in the environment needed before a command is executed.
+
+To avoid race conditions on reading and updating the status information for a command process, we're going to use a separate goroutine running `command.Wait()` to detect the moment when a process has finished and its status information has become available.
+
+##### Proposed library API
+
+The library is going to have a high-level abstraction for managing all commands within the system (a `ProcessManager`). It will be the main entry point for start/stop operations and for finding a command based on its command_id.
+
+Here are the proposed methods for the process manager:
+
+```go
+interface ProcessManager {
+  StartCommand(command []string) (*Command, error);
+  FindCommand(command_id string) *Command;
+  StopCommand(command_id) error;
+  Commands() []*Command;
+}
+```
+
+A lower level entity called `Command` will be used to wrap `exec.Command` and other Go APIs to implement a coherent API for managing a single process.
+
+```go
+interface Command {
+  Start() error;
+  Running() bool;
+  Wait();
+  Kill();
+  ResultCode() int32;
+  NewLogStream() *LogStream;
+  CloseLogStream(log *LogStream) error;
+}
+```
+
+Finally, a separate entity within the API will be used to implement a single log stream (`LogStream`). This should allow us to encapsulate all the streaming details within a single object with a simple API:
+
+```go
+interface LogStream {
+  NextLine() (log_line string, eof bool);
```

Comment:
What happens if the output does not contain any `\n`?

Reply:
Doh, this was based on my experiments before we updated the protocol. Will need to adjust it a bit to work with blocks of bytes, of course.

Reply:
Updated this. Now the caller could specify how much data they want to get (max limit) and how long they are willing to wait for it. This would make it possible for the caller to control how close to real time they want to get in their streaming code.